### ISSUE TITLE
proofsForTransactions in UtxoStateReader can modify the underlying tree

### ISSUE THREAD

**Issue by kushti** [ergoplatform/ergo#1387]
For the one node being stuck (with mining = true), the following was found in the log:

14:54:29.639 WARN  [tor.default-dispatcher-14] o.e.n.state.UtxoState - Error while applying full block with header cedc52078f826288f96d40f87c9992a44d1949c11c5ef316a8ae178a05231b5f to UTXOState with root 02154c9f004518a070ccab3a31e3ba0486fd3053fde09d15976372ff99c96d3c15, reason: java.lang.Error: Calculated proofHash is not equal to the declared one

Probably related to #1023 ( #1025 ), but initial investigation shows that likely an issue not in concurrent access to the underlying AVL+ (which should be prevented by synchronized access to persistent prover). Likely, BatchAVLProver.generateProofForOperations could (rarely) modify the underlying tree (or leave some internal structures dirty) in some edge cases.

**Comment by knizhnik** [comment]
I wonder how it is possible to reproduce the problem?
Just looking at the sources of BatchAVLProverI found one possible source of the problem:
resetNew method is intended to be used to "freeze" the tree: reset isNew flag in all nodes, so that any subsequent modification of the tree will do COW (copy on write). But recursion in this method is stopped if isNew is not set. But according to the AVL algorithm we may need to update leaves without updating upper internal nodes. In this case isNew flag will not be cleared. 

SO my  recommendation is to remove check for isNew in resetNew method:
```
diff --git a/src/main/scala/scorex/crypto/authds/avltree/batch/BatchAVLProver.scala b/src/main/scala/scorex/crypto/authds/avltree/batch/BatchAVLProver.scala
index a1505ef..fab1d59 100644
--- a/src/main/scala/scorex/crypto/authds/avltree/batch/BatchAVLProver.scala
+++ b/src/main/scala/scorex/crypto/authds/avltree/batch/BatchAVLProver.scala
@@ -296,16 +296,14 @@ class BatchAVLProver[D <: Digest, HF <: CryptographicHash[D]](val keyLength: Int
 
     // Recursively reset the new flags
     def resetNew(r: ProverNodes[D]): Unit = {
-      if (r.isNew) {
-        r match {
-          case rn: InternalProverNode[D] =>
-            resetNew(rn.left)
-            resetNew(rn.right)
-          case _ =>
-        }
-        r.isNew = false
-        r.visited = false
+      r match {
+        case rn: InternalProverNode[D] =>
+          resetNew(rn.left)
+          resetNew(rn.right)
+        case _ =>
       }
+      r.isNew = false
+      r.visited = false
     }
 
     packTree(oldTopNode)
```

**Comment by knizhnik** [comment]
I do not have permissions to push inthis repository so I can not create PR:
```
git push --set-upstream origin i1387
ERROR: Permission to input-output-hk/scrypto.git denied to knizhnik.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

**Comment by kushti** [comment]
@knizhnik invite sent.

On how to reproduce it, likely the best way is to run the test from #1023 and dump tree and operations leading to corrupted tree happen after proofsForOperations() , then make a test vector from the tree being dumped

**Comment by knizhnik** [comment]
There is still something strange with UtxoStateSpecification test.
If I perform more than one iteration of "concurrent applyModifier() and proofsForTransactions()" test (just wrap body of this test in a loop), then I always get error "Scripts of all transaction inputs should pass verification".
If I just increase number of iterations in parallel thread that generates proofs from 1000 till 100000, then test is passed, but some  subsequent tests are failed. Looks like this test doesn't wait termination of parallel thread.

**Comment by knizhnik** [comment]
Looks like the problem was not in resetNew.
I was able to reproduce "Interlinks should have the correct structure" error also with patched code.
Also I added assertion that there is no child with isNew=true, when parent has isNew=false. It never fired.

**Comment by knizhnik** [comment]
Hmmm....
I have already investigated the problem with interlinks, and there was PR for it:
https://github.com/ergoplatform/ergo/pull/1040
I wonder why it was not merged.

In any case the only problem I am reproducing at this moment is  "Interlinks should have the correct structure" caused by byte overflow. With fix  from PR #1040 the problem is not reproduced any more.

**Comment by abebeos** [comment]
@kushti , this issue is labeled https://github.com/ergoplatform/ergo/labels/P0-critical.

Why is it (and the related PR #1040) unprocessed?
